### PR TITLE
implement single sign-on flow

### PR DESCRIPTION
**Pull request by Rizziepit**

### CODE REVIEW

**Comment by Rizziepit** [comment]
@miltontony @smn ready for review.

This implements a minimal version of CAS 1.0. There is a Pyramid CAS client that we should be able to use with this.

File: unicore/hub/service/utils.py [hunk 1]
```diff
@@ -1,8 +1,63 @@
...
+    def get_active_policy(self, request):
+        path = request.path_info
+
+        if path not in self._path_info_cache:
+            self._path_info_cache[path] = self._resolve_policy(path)
+
+        return self._path_info_cache[path]
+
+    def authenticated_userid(self, request):
+        return self.get_active_policy(request) \
```
Comment (miltontony) [inline unicore/hub/service/utils.py]: A better style is to wrap lines on bracket and not use `\`

``` python
def authenticated_userid(self, request):
    return self.get_active_policy(
        request).authenticated_userid(request)
```

File: unicore/hub/service/utils.py [hunk 2]
```diff
@@ -1,8 +1,63 @@
...
+
 
 def make_password(bit_length=15):
     """ Returns a random password string
     """
     return base64.b64encode(os.urandom(bit_length))
+
+
+@implementer(IAuthenticationPolicy)
+class PathAuthenticationPolicy(object):
```
Comment (miltontony) [inline unicore/hub/service/utils.py]: It feels like it needs to be in a file called `policy.py` or `authentication.py` instead of utils.

**Comment by miltontony** [comment]
A bit of nitpicking, but :+1: overall

**Comment by smn** [comment]
Nice! Would it be worthwhile to release the CAS bits as `pyramid-mama-cas` ?

**Comment by Rizziepit** [comment]
@smn it could be, although I wouldn't call it `-mama-cas` because it implements the bare minimum - no CAS 2.0 or 3.0 features like `django-mama-cas`.

**Comment by smn** [comment]
ok cool.

On 12 March 2015 at 08:51, Rizmari Versfeld notifications@github.com
wrote:

> @smn https://github.com/smn it could be, although I wouldn't call it
> -mama-cas because it implements the bare minimum - no CAS 2.0 or 3.0
> features like django-mama-cas.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/universalcore/unicore.hub/pull/5#issuecomment-78431989
> .

## 

Simon de Haan | Chief Engineer Praekelt Foundation

praekeltfoundation.org | praekelt.com
M: +27 (0) 76 449 3806  |  http://twitter.com/smn

File: unicore/hub/service/utils.py (hunk 2 above)
Comment (Rizziepit) [inline unicore/hub/service/utils.py]: Done

File: unicore/hub/service/utils.py (hunk 1 above)
Comment (Rizziepit) [inline unicore/hub/service/utils.py]: Done

**Comment by miltontony** [comment]
:+1:

**Pull request merged**